system/lib/emmalloc.cpp

```diff
@@ -0,0 +1,1187 @@
+/*
+ * Simple minimalistic but efficient malloc/free.
+ *
+ * Assumptions:
+ *
+ *  - 32-bit system.
+ *  - Single-threaded.
+ *  - sbrk() is used, and nothing else.
+ *  - sbrk() will not be accessed by anyone else.
+ *  - sbrk() is very fast in most cases (internal wasm call).
+ *
+ * Invariants:
+ *
+ *  - Metadata is 8 bytes, allocation payload is a
+ *    multiple of 8 bytes.
+ *  - All regions of memory are adjacent.
+ *  - Due to the above, after initial alignment fixing, all
+ *    regions are aligned.
+ *  - A region is either in use (used payload > 0) or not.
+ *    Used regions may be adjacent, and a used and unused region
+ *    may be adjacent, but not two unused ones - they would be
+ *    merged.
+ *  - A used region always has minimal space at the end - we
+ *    split off extra space when possible immediately.
+ *
+ * Debugging:
+ *
+ *  - If not NDEBUG, runtime assert()s are in use.
+ *  - If EMMALLOC_DEBUG is defined, a large amount of extra checks are done.
+ *  - If EMMALLOC_DEBUG_LOG is defined, a lot of operations are logged
+ *    out, in addition to EMMALLOC_DEBUG.
+ *  - Debugging and logging uses EM_ASM, not printf etc., to minimize any
+ *    risk of debugging or logging depending on malloc.
+ *
+ * TODO
+ *
+ *  - Optimizations for small allocations that are not multiples of 8, like
+ *    12 and 20 (which take 24 and 32 bytes respectively)
+ *
+ */
+
+#include <assert.h>
+#include <malloc.h> // mallinfo
+#include <string.h> // for memcpy, memset
+#include <unistd.h> // for sbrk()
+#include <emscripten.h>
+
+#define EMMALLOC_EXPORT __attribute__((__weak__, __visibility__("default")))
+
+// Debugging
+
+#ifdef EMMALLOC_DEBUG_LOG
+#ifndef EMMALLOC_DEBUG
+#define EMMALLOC_DEBUG
+#endif
+#endif
+
+#ifdef EMMALLOC_DEBUG
+// Forward declaration for convenience.
+static void emmalloc_validate_all();
+#endif
+#ifdef EMMALLOC_DEBUG
+// Forward declaration for convenience.
+static void emmalloc_dump_all();
+#endif
+
+// Math utilities
+
+static bool isPowerOf2(size_t x) {
+  return __builtin_popcount(x) == 1;
+}
+
+static size_t lowerBoundPowerOf2(size_t x) {
+  if (x == 0) return 1;
+  // e.g. 5 is 0..0101, so clz is 29, and we want
+  // 4 which is 1 << 2, so the result should be 2
+  return 31 - __builtin_clz(x);
+}
+
+// Constants
+
+// All allocations are aligned to this value.
+static const size_t ALIGNMENT = 8;
+
+// Even allocating 1 byte incurs this much actual payload
```

Comment:
Will allocating 0 bytes just allocate the 8 byte metadata, or also 8 bytes of payload?

Reply:
Allocating 0 just returns NULL, actually, which is allowed by the [docs](https://linux.die.net/man/3/malloc)
```
 If size is 0, then malloc() returns either NULL, or a unique pointer value that can later be successfully passed to free(). 
```